lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -376,6 +376,51 @@ class DecomposeAtenAddmmOp : public OpRewritePattern<AtenAddmmOp> {
 };
 } // namespace
 
+// Decompose torch.mean into: sum(x)/div(numTensorElements).
+namespace {
+class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenMeanOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.self();
+    Value output = op.result();
+
+    Value dtype = rewriter.create<ConstantNoneOp>(loc);
+    Value sum = rewriter.create<AtenSumOp>(loc, output.getType(), input, dtype);
+    if (!sum)
+      return failure();
+
+    BaseTensorType tensorType = input.getType().cast<BaseTensorType>();
+    SmallVector<int64_t> sizes;
+    int64_t tensorSize = 1;
+    if (tensorType.hasSizes()) {
+      ArrayRef<int64_t> inputShape = tensorType.getSizes();
+      for (int64_t i : inputShape) {
+        if (i < 1)
+          return rewriter.notifyMatchFailure(op, "Invalid tensor size.");
+        tensorSize *= i;
+      }
+    } else
+      return rewriter.notifyMatchFailure(
+          op, "Only support input with known sizes.");
+
+    Value constTensorSize = rewriter.create<ConstantIntOp>(
+        loc, rewriter.getI64IntegerAttr(tensorSize));
+    Type divisorType =
+        tensorType.getWithSizesAndDtype({}, Torch::IntType());
```

Comment:
I think there is an issue with `getWithSizesAndDtype` here. the second argument dtype should be builtin types see https://github.com/dan-garvey/torch-mlir/blob/4cf554704cd1c417d6f71f53b107df062d0fe331/lib/Dialect/Torch/IR/TorchTypes.cpp#L53.  If the type is specified correctly here, the op should have <[],si64> as expected.

We don't put this before refine types because some decompositions depends on knowing the rank of the tensor. 
Currently, the ops created in `DecomposeComplexOps.cpp` need to take care of specifying accurate types for the newly generated ops because it's after refinetypes. 

In the future, we will see if we need to split the ops into the ones don't need refined type info and ones do.

Reply:
That makes more sense

Reply:
@cathyzhyi , Is this always true for `getWithSizesAndDtype`? In particular, should the function check if the type is a valid dtype by using 

https://github.com/llvm/torch-mlir/blob/ea7a30f9b9b14f469024db35ce16bea2a5ee9548/lib/Dialect/Torch/IR/TorchTypes.cpp#L53

Reply:
yea, I think adding an assert at entry of `getWithSizesAndDtype` is a good idea. Even without the assert, I would expect this to be caught at this line. https://github.com/llvm/torch-mlir/blob/127c7d8e27ede211af77392b5277ba06fb515fb7/lib/Dialect/Torch/IR/TorchTypes.cpp#L172. It's strange that with an invalid dtype passed in the `getWithSizesAndDtype` just returns a ` !torch.vtensor`.

Reply:
I think I know the issue. I'm not sure why, but the constructor for `Torch::IntType` returns a null. I think this is because the actual constructor for the type is `Torch::IntType::get(...)`. Since the verifier ignores null types, then it does not raise an error.

Reply:
Yeah, I just verified this. If you change `Torch::IntType()` to `Torch::IntType::get(op.getContext())` you get the following error:

```
unknown>:0: error: invalid dtype '!torch.int' for !torch.tensor type
torch-mlir-opt: ../external/llvm-project/llvm/../mlir/include/mlir/IR/StorageUniquerSupport.h:140: static ConcreteT mlir::detail::StorageUserBase<mlir::torch::Torch::ValueTensorType, mlir::torch::Torch::BaseTensorType, mlir::torch::Torch::detail::ValueTensorTypeStorage, mlir::detail::TypeUniquer>::get(mlir::MLIRContext *, Args...) [ConcreteT = mlir::torch::Torch::ValueTensorType, BaseT = mlir::torch::Torch::BaseTensorType, StorageT = mlir::torch::Torch::detail::ValueTensorTypeStorage, UniquerT = mlir::detail::TypeUniquer, Traits = <>, Args = <llvm::Optional<llvm::ArrayRef<long>>, mlir::Type>]: Assertion `succeeded(ConcreteT::verify(getDefaultDiagnosticEmitFn(ctx), args...))' failed.
```

Reply:
Cool! Thanks for digging so deep into this!

Reply:
Thanks, @ramiro050 and @cathyzhyi, for your comments and suggestions. I am closing this pull request as a push from the forked repo created this.